sdk/src/mesh_dp.c

```diff
@@ -0,0 +1,779 @@
+
+/**
+ * Mesh client context structure
+ */
+typedef struct MeshClientContext {
+    MeshClientConfig config;
+
+    struct MeshConnectionListHead conn_list_head;
+    pthread_mutex_t mx;
+
+    atomic_int conn_num;
+
+} MeshClientContext;
+
+/**
+ * Mesh connection buffer structure
+ */
+typedef struct MeshBufferContext {
+    /**
+     * NOTE: The __public structure is directly mapped in the memory to the
+     * MeshBuffer structure, which is publicly accessible to the user.
+     * Therefore, the __public structure _MUST_ be placed first here.
+     */
+    MeshBuffer __public;
+
+    /**
+     * NOTE: All declarations below this point are hidden from the user.
+     */
+    mcm_buffer *buf;
+
+} MeshBufferContext;
+
+/**
+ * Create a new mesh client
+ */
+int mesh_create_client(MeshClient **mc, MeshClientConfig *cfg)
+{
+    MeshClientContext *mc_ctx;
+    int err;
+
+    if (!mc)
+        return -MESH_ERR_BAD_CLIENT_PTR;
+
+    mc_ctx = mesh_calloc(sizeof(MeshClientContext));
+    if (!mc_ctx) {
+        *mc = NULL;
+        return -ENOMEM;
+    }
+
+    atomic_init(&mc_ctx->conn_num, 0);
+
+    err = pthread_mutex_init(&mc_ctx->mx, NULL);
+    if (err) {
+        mesh_free(mc_ctx);
+        *mc = NULL;
+        return err;
+    }
+
+    LIST_INIT(&mc_ctx->conn_list_head);
+
+    if (cfg)
+        mesh_memcpy(&mc_ctx->config, cfg, sizeof(MeshClientConfig));
+
+    if (mc_ctx->config.max_conn_num == 0)
+        mc_ctx->config.max_conn_num = MESH_CLIENT_DEFAULT_MAX_CONN;
+
+    if (mc_ctx->config.timeout_ms == 0)
+        mc_ctx->config.timeout_ms = MESH_CLIENT_DEFAULT_TIMEOUT_MS;
+
+    *mc = (MeshClient *)mc_ctx;
+
+    return 0;
+}
+
+/**
+ * Delete mesh client
+ */
+int mesh_delete_client(MeshClient **mc)
+{
+    MeshClientContext *mc_ctx;
+
+    if (!mc)
+        return -MESH_ERR_BAD_CLIENT_PTR;
+
+    mc_ctx = (MeshClientContext *)(*mc);
+
+    if (!mc_ctx)
+        return -MESH_ERR_BAD_CLIENT_PTR;
+
+    pthread_mutex_destroy(&mc_ctx->mx);
+
+    /**
+     * TODO: Shutdown and deallocate connections here
+     */
+
+    mesh_free(mc_ctx);
+    *mc = NULL;
+
+    return 0;
+}
+
+/**
+ * Create a new mesh connection
+ */
+int mesh_create_connection(MeshClient *mc, MeshConnection **conn)
+{
+    MeshConnectionContext *conn_ctx;
+    MeshClientContext *mc_ctx;
+    int conn_num, err;
+
+    if (!mc)
+        return -MESH_ERR_BAD_CLIENT_PTR;
+
+    if (!conn)
+        return -MESH_ERR_BAD_CONN_PTR;
+
+    mc_ctx = (MeshClientContext *)mc;
+
+    conn_num = atomic_load_explicit(&mc_ctx->conn_num, memory_order_relaxed);
+    while (conn_num < mc_ctx->config.max_conn_num) {
+        if (atomic_compare_exchange_weak_explicit(&mc_ctx->conn_num, &conn_num,
+                                                  conn_num + 1,
+                                                  memory_order_release,
+                                                  memory_order_relaxed))
+            break;
+    }
+    if (!(conn_num < mc_ctx->config.max_conn_num))
```

Comment:
```suggestion
    if ((conn_num > mc_ctx->config.max_conn_num))
```
I think it is alright when `conn_num  == mc_ctx->config.max_conn_num`. We just don't want to have more then max.